### PR TITLE
A Login Frontend made using React.js using Django-Rest Framework APIs as backend

### PR DESCRIPTION
**Pull request by musanaeem**
I created a Front-end mechanism for user login using API calls to the already made Django-Rest Framework. Authentication is done in the backend and the front ends job is to ensure user is logged in, store tokens and reroute user accordingly. 

- I made use of class components for this task and styled the frontend using a combination of bootstrap and css. 
- Since React doesn't allow a CSRF token tag, I made a csrftoken.js file that gets the token for me and I render it in the form.
- The Login functionality was made using controlled components - where event calls are made to the parent component -and normal components.
- ProtectedRoutes.js and ProtectedLogin.js are responsible for rerouting user according to if they are logged in or not.
- Routing is the parent class that Routes and renders the page components and handles events.
- Function Wrappers e.g RoutingWrapper are functions made in order to make use of hooks that are not compatible with class components. wrapper functions were used as a temporary workaround before conversion to functional components.
- Pages that are to be rendered and shown to user are put underneath the pages directory and tools to be used to make up those pages are put under components.
- The Helmet tag from "react-helmet" is used t to access the header tag to add the styling links.
- Bootstrap is installed in the react app.
- Finally jwt tokens are used to set cookies and if the cookie is set, the user can access pages otherwise the user is sent back to the login page. If the jwt token is set, the user can not go to login page. The logout button removes this token. Token authentication is done at backend so even if a random token is set, the user can not query data as they will not be authenticated.

Screenshots:

Login Page:
image

Placeholder HomePage:
image

User does not exist error:
image

Incorrect password error:
image

### CODE REVIEW

File: src/components/LoginForm.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import React, { Component } from 'react'
+import CSRFToken from './csrftoken'
```
Comment (srehaider) [inline src/components/LoginForm.js]: Not required anymore.

File: src/components/Routing.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      },
+      body: JSON.stringify(credentials),
+    };
+
+    return options
+  }
+  
+
+
+  logUserIn = credentials => {
```
Comment (srehaider) [inline src/components/Routing.js]: Move it to a separate service.

File: src/components/Routing.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    
+    let options = this.getOptions(credentials)
+
+
+    fetch('http://localhost:8000/api/login/', options)
```
Comment (srehaider) [inline src/components/Routing.js]: Get API domain from environment variables(`.env`)

**Review by srehaider — APPROVED** [review]

File: src/components/LoginForm.js (hunk 1 above)
Comment (musanaeem) [reply src/components/LoginForm.js]: Removed.

File: src/components/Routing.js (hunk 2 above)
Comment (musanaeem) [reply src/components/Routing.js]: Done.

File: src/components/Routing.js (hunk 3 above)
Comment (musanaeem) [reply src/components/Routing.js]: done

File: src/pages/LoginPage.js [hunk 4]
```diff
@@ -2,8 +2,38 @@ import React, { Component } from 'react'
...
+class LoginPage extends Component {
+
+    constructor(props){
+        super(props);
+
+        this.state = {
+            error: ''
+        }
+
+        this.logUserIn = this.logUserIn.bind();
```
Comment (srehaider) [inline src/pages/LoginPage.js]: You don't need to bind it, it's an arrow function it by default should have class context.
```suggestion
```

File: src/services/LoginRequest.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import Cookies from 'js-cookie';
+
+async function LoginRequest(credentials) {
```
Comment (srehaider) [inline src/services/LoginRequest.js]: Only class names start with capital letters and you are not using `await` in function no need to make it `async` 
```suggestion
function loginRequest(credentials) {
```

File: src/services/LoginRequest.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+            'Authorization': null,
+            'X-CSRFToken': Cookies.get('csrftoken')
+          },
+          body: JSON.stringify(credentials),
+        };
+    
+        return options
+      }
```
Comment (srehaider) [inline src/services/LoginRequest.js]: Extract this function outside the block of `loginRequest` and fix the indentation.

**Review by srehaider — APPROVED** [review]

**Pull request merged**